libcloud/compute/drivers/packet.py

```diff
@@ -88,11 +98,88 @@ class PacketNodeDriver(NodeDriver):
                       'failed': NodeState.ERROR,
                       'active': NodeState.RUNNING}
 
-    def list_nodes(self, ex_project_id):
-        data = self.connection.request('/projects/%s/devices' %
-                                       (ex_project_id),
-                                       params={'include': 'plan'}
-                                       ).object['devices']
+    def __init__(self, key, project=None):
+        # initialize a NodeDriver for Packet using the API token
+        # and optionally the project (name or id)
+        # If project specified we need to be sure this is a valid project
+        # so we create the variable self.project_id
+        super(PacketNodeDriver, self).__init__(key=key, project=None)
+        self.project_name = project
+        self.project_id = None
+        self.projects = self.ex_list_projects()
```

Comment:
Performing any kind of long running operations inside the constructor is not the best thing to do.

One approach would be to run this code lazily during the first HTTP request if ``self.project_id`` or ``self.project_name`` variable is not set yet.

In addition to that, there could be some kind of argument / method which user could use to force check if a specified project is valid.

Reply:
Agreed. Are there examples of such lazy requests in other drivers?

Reply:
Can't recall from top of my head.

Can probably just use @property and internal variable which is set to ``None`` inside the constructor and inside the property method we call the API if that value is none. Something along the lines of:

```python
@property
def projects(self):
    if not self._projects:
        self._projects = self.ex_list_projects()

    return self._projects
```

Reply:
I believe something like this should do it - https://github.com/Kami/libcloud/commit/8e2e9b18d0db61cb16947ce00f227c76e944f13c, but I didn't get a chance to test it with actual Packet.net API yet.

As you can see, I added additional guard condition to avoid infinite loop in case account has no projects.